refactor: aggregate duplicated code in servicer

airkei opened the pull request:
### Why
The motivation is to add client update command. Before that, organize and remove duplicated code in servicer.

### What
Aggregate the duplicated code for update and rollback in servicer.

### Tests
Confirmed to pass the unit tests.

Bodong-Yang commented:
@airkei The idea is greate, but the implementation needs to be polished. Please consider how to properly type the newly added aggregated functions :pray:

Bodong-Yang commented:
Suggest that instead of providing `request_cls, local_handler, request_cls, remote_call, response_type, response_type, response_ecu_type` as args, what about just checking the type of `request`, and let the `_handle_request` decides what to do.

Something like:
```python
  def _handle_request(self, request: api_types.UpdateRequest | api_types.RollbackRequest):
        if isinstance(request, api_types.UpdateRequest):
            local_handler=self._local_update,
            request_cls=UpdateRequestV2,
            remote_call=OTAClientCall.update_call,
            response_type=api_types.UpdateResponse,
            response_ecu_type=api_types.UpdateResponseEcu,
        else:
            ...
```

By this the code will be simple and readability will be improved.

airkei commented:
@Bodong-Yang Thank you for your feedback.
I understood how it is difficult to aggregate the methods of the different inputs/outputs in Python... Especially in the combination of static code analysis and dynamic typing language :disappointed:

In the latest commit, introduced fixes.
- set `update_acked_ecus` only for `update`. It means main ECU will not wait for others in `rollback`.
- added overload and type hint for each function
- for overload and type hint, let me pass `request_cls` as arg instead of judging from `request`.

airkei commented:
Added test file for servicer.
- update
Verified that `on_ecus_accept_update_request` is called.
https://github.com/tier4/ota-client/blob/12377ae5a32fd6a5b29bcb47121ce18fd4cb3dda/tests/test_otaclient/test_grpc/test_api_v2/test_servicer.py#L260

- rollback
Verified that `on_ecus_accept_update_request` is not called.
https://github.com/tier4/ota-client/blob/12377ae5a32fd6a5b29bcb47121ce18fd4cb3dda/tests/test_otaclient/test_grpc/test_api_v2/test_servicer.py#L435

Bodong-Yang commented:
> I understood how it is difficult to aggregate the methods of the different inputs/outputs in Python... Especially in the combination of static code analysis and dynamic typing language 😞

@airkei 

This time the typing is hard due to:
1. Container message like `UpdateResponse` and its contained type `UpdateResponseEcu` currently are not related to each other from python typing point of view. Extra typings are needed for these two types(like `list[str]`, announcing that the container type has `str` as element type).
2. `UpdateResponse` and `RollbackResponse`, they behave similarly (both have same helper methods), but these helper methods takes different contained type, so we still need to narrow down the types before we can typing the methods call.
3. I also think of using `overload` to type the aggregated function(`_handle_request`), but the limitation is it only checks the input of function call, inside the function itself we still need to manually reveal the type(with `isinstance`) to make type checker happy.

You have already done a good job, changes of PR is the best we can do to the typing for now.

Bodong-Yang commented:
( Also due to how python typing works, actually providing `request_cls, local_handler, request_cls, remote_call, response_type, response_type, response_ecu_type` as args has no difference to using `isinstance` over `request` and at runtime from typing point of view. 
For both case, inside `_handle_request`, you still need to reveal the type from Union.

airkei commented:
@Bodong-Yang Thank you for your feedback! I also tried to pass only `request` as arg and use `isinstance` internally based on your adivce. In this case, caller side complains for the ambiguous **output** :cry:
branch: https://github.com/tier4/ota-client/commit/d25f817840f338b342e1864ebac2471212b78f9b
```
Type "UpdateResponse | RollbackResponse" is not assignable to return type "UpdateResponse"
  Type "UpdateResponse | RollbackResponse" is not assignable to type "UpdateResponse"
    "RollbackResponse" is not assignable to "UpdateResponse"Pylance[reportReturnType](https://github.com/microsoft/pyright/blob/main/docs/configuration.md#reportReturnType)
(method) def _handle_request(request: UpdateRequest | RollbackRequest) -> CoroutineType[Any, Any, UpdateResponse | RollbackResponse]
```

![image](https://github.com/user-attachments/assets/41b09d2c-0425-4214-809e-cb288dcc730d)

Bodong-Yang commented:
>  caller side complains for the ambiguous output 😢
@airkei  Oh since we type `_handle_request` return value as Union, we need to narrow it down to one of the type:

```python
async def update(self, request: api_types.UpdateRequest) -> api_types.UpdateResponse:
    # here we know that for update request, we will get UpdateResponse from _handle_request, so type hint the res as UpdateResponse to narrow down the Union to concrete type
    res: api_types.UpdateResponse = await self._handle_request(request=request)
    return res
```
